torch/utils/_content_store.py

```diff
@@ -0,0 +1,156 @@
+# portable across devices) that is NOT cryptographically secure.  In return,
+# we are able to hash 40G of tensor data on GPU in less than a second,
+# compared to running SHA-1 in CPU which would a minute or so.  The primary
+# use case is for efficiently snapshotting intermediate tensor data for
+# offline debugging, but it's been put in this module in case you think of
+# another use case for it.  The hash function could be replaced with a
+# straight reimplementation of SHA-1, which would give us much stronger
+# portability guarantees.
+#
+# WARNING: THERE IS NO BC/FC GUARANTEE FOR THIS FORMAT!  If you need to format
+# shift the result, consider packing it into a single torch.save object
+# with traditional view sharing.
+#
+# Because of the weak portability guarantees, you can only write to the
+# content store from a single process; we don't provide any capability
+# of "reopening" a content store to add more things to it.  But we don't
+# assume that you can keep all of the tensors you want to add to the store
+# in memory at once, because you probably can't!  Nor do we assume that
+# you know a priori whether or not two storages can be deduplicated or not.
+#
+# Note: only storages are content-addressed; tensors are name addressed
+#
+# Note: our padding strategy means that [1, 0] and [1] int16 tensors would
+# map to the same (padded) storage.  We think this will be immaterial for most
+# users.
+
+import os.path
+import struct
+from typing import Optional
+
+import torch
+import torch._prims as prims
+import torch._utils
+import torch.nn.functional as F
+
+from torch.multiprocessing.reductions import StorageWeakRef
+
+
+def hash_storage(storage):
+    # Use of torch.compile is mandatory for (1) good memory usage
+    # and (2) xor_sum implementation
+    @torch.compile(dynamic=True)
+    def kernel(x):
+        # The randint calls are carefully written to hit things we
+        # have lowerings for in inductor.  Lack of unsigned 32-bit integer
+        # is a pain.
+        a = torch.randint(
+            -(2**31), 2**31, x.shape, device=x.device, dtype=torch.int32
+        ).abs()
+        a = ((a % (2**31 - 1)) + 1).long()
+        b = (
+            torch.randint(
+                -(2**31), 2**31, x.shape, device=x.device, dtype=torch.int32
+            )
+            .abs()
+            .long()
+        )
+        # This is a standard shift-multiply universal hash family
+        # plus xor sum hash, using Philox to generate random numbers.
+        # Our Philox RNG is not deterministic across devices so
+        # don't use this for stable hashing.
+        #
+        # This assumes fixed length so you're also obligated to bucket
+        # by the length of tensor as well
+        return prims.xor_sum((a * x + b).int(), [0])
+
+    with torch.random.fork_rng(
+        [storage.device] if storage.device.type != "cpu" else []
+    ):
+        torch.manual_seed(0)  # this can be anything, just needs to be fixed
```

Comment:
Will this change manual seed for the rest of the program? Do you need to set it back?

Reply:
OK so apparently the random api is full of footguns. Rewrite is much better.